service/history/handler.go

```diff
@@ -197,11 +196,32 @@ func (h *Handler) isStopped() bool {
 	return atomic.LoadInt32(&h.status) == common.DaemonStatusStopped
 }
 
-// IsWorkflowTaskValid - whether workflow task is still valid
-func (h *Handler) IsWorkflowTaskValid(ctx context.Context, request *historyservice.IsWorkflowTaskValidRequest) (_ *historyservice.IsWorkflowTaskValidResponse, retError error) {
-	defer log.CapturePanic(h.logger, &retError)
+func lifecycleInterceptor(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (_ interface{}, retErr error) {
+	h, _ := info.Server.(*Handler)
+
+	// wait until service is ready to serve any requests
 	h.startWG.Wait()
```

Comment:
Is it possible for this to never return if we shut down too early?

Reply:
From what I can tell, this should only block during the start of the server, not the shutdown. Once `Start()` completed, the `.Wait()` will always be unblocked.